Agents/BMSUpdateAgent/BMSUpdateAgent/src/main/java/uk/ac/cam/cares/jps/agent/bmsupdate/BMSUpdateAgent.java

```diff
@@ -144,4 +164,225 @@ private String parseFanStatusMessage(String response) {
+     *
+     * @param dataIRI The data iri
+     * @param rsClient The knowledge graph client
+     * @return Bacnet Device ID of data IRI
+     */
+    public String getDeviceId(String dataIRI, RemoteStoreClient rsClient) {
+        SelectQuery selectQuery = Queries.SELECT();
+        Variable deviceIdVar = SparqlBuilder.var("deviceIdVar");
+        TriplePattern getBacnetDeviceId = GraphPatterns.tp(iri(dataIRI), P_ONTOBMS.iri("hasBacnetDeviceID"), deviceIdVar);
+        selectQuery.prefix(P_ONTOBMS).select(deviceIdVar).where(getBacnetDeviceId);
+        try {
+            LOGGER.info("getting Bacnet Device ID of " + dataIRI +" ...");
+            LOGGER.info("execute query: " + selectQuery.getQueryString());
+            JSONArray results = new JSONArray(rsClient.execute(selectQuery.getQueryString()));
+            return results.getJSONObject(0).getString("deviceIdVar");
+        } catch (Exception e) {
+            LOGGER.error(FAIL_TO_GET_BACNETDEVICEID);
+            throw new JPSRuntimeException(FAIL_TO_GET_BACNETDEVICEID);
+        }
+    }
+
+    /**
+     * Get the Bacnet Object ID of the data IRI
+     *
+     * @param dataIRI The data iri
+     * @param rsClient The knowledge graph client
+     * @return Bacnet Object ID of data IRI
+     */
+    public String getObjectId(String dataIRI, RemoteStoreClient rsClient) {
+        SelectQuery selectQuery = Queries.SELECT();
+        Variable objectIdVar = SparqlBuilder.var("objectIdVar");
+        TriplePattern getBacnetObjectId = GraphPatterns.tp(iri(dataIRI), P_ONTOBMS.iri("hasBacnetObjectID"), objectIdVar);
+        selectQuery.prefix(P_ONTOBMS).select(objectIdVar).where(getBacnetObjectId);
+        try {
+            LOGGER.info("getting Bacnet Object ID of " + dataIRI +" ...");
+            LOGGER.info("execute query: " + selectQuery.getQueryString());
+            JSONArray results = new JSONArray(rsClient.execute(selectQuery.getQueryString()));
+            return results.getJSONObject(0).getString("objectIdVar");
+        } catch (Exception e) {
+            LOGGER.error(FAIL_TO_GET_BACNETOBJECTID);
+            throw new JPSRuntimeException(FAIL_TO_GET_BACNETOBJECTID);
+        }
+    }
+
+    /**
+     * Get latest timeseries value for the data IRI and compared it with the trigger value,
+     * if they are equal, run a sparql update to insert a set of triples and delete a set of triples
+     * @param dataIRI The data iri
+     * @param rsClient The knowledge graph client
+     * @param RDBClient The remote RDB client
+     * @param triggerValue The value to compare with the latest timeseries value
+     * @param insertString The string to insert via sparql update
+     * @param deleteString The string to delete via sparql update
+     * @return 
+     */
+    public JSONObject checkInsertAndDelete(String dataIRI, RemoteStoreClient rsClient, RemoteRDBStoreClient RDBClient, String triggerValue, String insertString, String deleteString) {
```

Comment:
Code duplication. Same pieces of code appear in checkInsert() and checkDelete(). Better to split the function to three private base functions check(), inert() and delete() and call these base functions in checkInsertAndDelete(), checkInsert() and checkDelete().

Reply:
Refactored code based on your suggestion. Now there is a check(), insert() and delete(). The checkInsertAndDelete(), checkInsert() and checkDelete() calls these methods